syntax = "proto3";

option java_package = "ex.grpc";
option objc_class_prefix = "RTG";

package sankalpafs;

// Interface exported by the server.
service SankalpaFS {
  rpc get_mtime(Path) returns (MTime) {}

  rpc get_file_contents(Path) returns (stream Content) {}

  rpc update_file(stream Content) returns (UpdateAck) {}

  rpc delete(Path) returns (Status) {}

  rpc mkdir(Path) returns (Status) {}

  rpc rmdir(Path) returns (Status) {}

  rpc readdir(Path) returns (ListDir) {}

  rpc getattr(Path) returns (Stat) {}

  rpc rename(SrcDst) returns (Status) {}

}

message Path {
  string path = 1;
}

message MTime {
  double mtime = 1;
}

message Content {
  string content = 1;
}

message UpdateAck {
  string file_path = 1;
  int64 num_bytes = 2;
  MTime server_mtime = 3;
}

message Status {
  int32 status = 1;
}

message ListDir {
  int32 status = 1;
  repeated string dir = 2;
}

message Stat {
  uint32 st_mode = 1;
  uint32 st_ino = 2;
  uint32 st_dev = 3;
  uint32 st_nlink = 4;
  uint32 st_uid = 5;
  uint32 st_gid = 6;
  uint32 st_size = 7;
  double st_atime = 8;
  double st_mtime = 9;
  double st_ctime  = 10;
  int32 status = 11;
}

message SrcDst {
  string src = 1;
  string dst = 2;
}